clients/fides-js/src/lib/i18n/i18n-utils.ts

```diff
@@ -0,0 +1,357 @@
+ * 3) Fallback to the {language} only if possible (e.g. "fr-CA" -> "fr")
+ * 4) Fallback to the default locale otherwise (e.g. "fr-CA" -> "en")
+ */
+export function matchAvailableLocales(
+  requestedLocale: Locale,
+  availableLocales: Locale[]
+): Locale {
+  // 1) Parse the requested locale string using our regex
+  const match = requestedLocale.match(LOCALE_REGEX);
+  if (match) {
+    const [locale, language] = match;
+
+    // 2) Look for an exact match of the requested locale
+    const exactMatch = availableLocales.find(
+      (elem) =>
+        elem.toLowerCase().replaceAll("_", "-") ===
+        locale.toLowerCase().replaceAll("_", "-")
+    );
+    if (exactMatch) {
+      return exactMatch;
+    }
+
+    // 3) Fallback to the {language} of the requested locale
+    const languageMatch = availableLocales.find(
+      (elem) =>
+        elem.toLowerCase().replaceAll("_", "-") ===
+        language.toLowerCase().replaceAll("_", "-")
+    );
+    if (languageMatch) {
+      return languageMatch;
+    }
+  }
+
+  // 4) Fallback to default locale
+  return DEFAULT_LOCALE;
+}
+
+/**
+ * Initialize the given i18n singleton by:
+ * 1) Loading all static messages from locale files
+ * 2) Detecting the user's locale
+ * 3) Activating the best match for the user's locale
+ */
+export function initializeI18n(
+  i18n: I18n,
+  navigator: Partial<Navigator>,
+  experience: Partial<PrivacyExperience>,
+  options?: Partial<FidesOptions>
+): void {
+  // Extract & update all the translated messages from both our static files and the experience API
+  updateMessagesFromFiles(i18n);
+  const availableLocales = updateMessagesFromExperience(i18n, experience);
+
+  // Detect the user's locale, unless it's been *explicitly* disabled in the experience config
+  let userLocale = DEFAULT_LOCALE;
+  // TODO: update types and remove any
+  if (
+    !((experience as any)?.experience_config?.auto_detect_language === false)
+  ) {
+    userLocale = detectUserLocale(navigator, options);
+  }
+
+  // Match the user locale to the "best" available locale from the experience API and activate it!
+  const bestLocale = matchAvailableLocales(userLocale, availableLocales);
+  i18n.activate(bestLocale);
+}
+
+/**
+ * Factory function that returns a new I18n instance that is the *simplest*
+ * possible implementation of an i18n library that meets our minimal
+ * requirements as of February 2024.
+ *
+ * Note that this implementation is designed to be completely swapped out for
+ * LinguiJS once we're ready to upgrade to the real thing!
+ */
+export function setupI18n(): I18n {
+  // Currently active locale; default this to English
+  let currentLocale: Locale = DEFAULT_LOCALE;
+
+  // Messages catalog, which stores i18n-ized messages grouped by locale
+  const allMessages: Record<Locale, Messages> = {};
+
+  // Return a new I18n instance
+  return {
+    activate: (locale: Locale): void => {
+      currentLocale = locale;
+    },
```

Comment:
When we implement language switching from the UI, all we'll need to do is call this `activate` fn with the chosen locale, right?

Reply:
Yup 👍